File: bin/gpa2xml.rb
#!/usr/bin/env ruby

# Takes a Go PlayAlong XML file with synchronization data and Guitar Pro tab
# and produces Rocksmith 2014 XML tracks.

# Put sync points on signature change ! Use expand repeats !
# Section are markers
# Phrase are double bars
# Anchors are determined by fingering
# Copyright => Year
# AlbumArt =>
# Tones => Note.text defines tone name and tone change
# base tone is 'ToneBase'
# cent offset ??

require 'rexml/document'
require 'interpolator'
require 'guitar_pro_parser'
require 'gyoku'
require 'matrix'

# To be removed since it already exists in master for guitar_pro_parser
def GuitarProHelper.note_to_digit(note)
  result = 0
  result += 1 until GuitarProHelper.digit_to_note(result) == note
  result
end

## HELPERS

TONEBASE = 'ToneBase'

def time_interpolator(sync)
  s = sync.split '#'
  s.shift

  t = s.map { |u| u.split ';' }

  m = t.map do |time, bar, bar_fraction, beat_duration|
    { bar.to_f + bar_fraction.to_f => time.to_f / 1000.0 }
  end

  Interpolator::Table.new m.reduce Hash.new, :merge
end

I_DURATIONS = GuitarProHelper::DURATIONS.invert

STANDARD_TUNING = %w(E3 A3 D4 G4 B4 E5).map do |n|
  GuitarProHelper.note_to_digit(n)
end

def sortable_name(n)
  n.sub(/^(the|a|an)\s+/i, '').capitalize
end

# Counted arrays
def carray(symbol, arr)
  {
    :@count => arr.count,
    :content! => { symbol => arr }
  }
end

def compute_tuning(gp_tuning)
  t = gp_tuning.reverse.map { |n| GuitarProHelper.note_to_digit(n) }
  x = Vector.elements(t) - Vector.elements(STANDARD_TUNING)

  {
    :@string0 => x[0], :@string1 => x[1], :@string2 => x[2],
    :@string3 => x[3], :@string4 => x[4], :@string5 => x[5]
  }
end

## MAIN STUFF

# TODO: add time offset ?
# Build notes and chords, mostly effects
# anchors
# handshapes
# chord templates
# fingering
# arrangement properties
# section played number
# alternate endings ?

class SngXmlBuilder
  def initialize(song, track, timefun)
    @song = song
    @track = track
    @timerinter = timefun

    @internal_name = @song.artist.gsub(/[^0-9a-z]/i, '') + '_'
    @internal_name += @song.title.gsub(/[^0-9a-z]/i, '')

    @signature = 4.0 # 4/4 default signature

    @notes = []
    @chords = []
    @ebeats = []
    @sections = []
    @tones = []
    @tone_map = [TONEBASE]

    @phrases = [
      {
        :@disparity => 0,
        :@name => '',
        :@ignore => 0,
        :@maxDifficulty => 0,
        :@solo => 0
      }
    ]

    @phrase_iterations = [
      {
        :@time => 10.000,
        :@phraseId => 0,
        :@variation => ''
      }
    ]

    @chord_templates = [
      {
        :@chordName => '',
        :@displayName => 'g5p5',
        :@finger0 => -1,
        :@finger1 => -1,
        :@finger2 => -1,
        :@finger3 => -1,
        :@finger4 => 1,
        :@finger5 => 1,
        :@fret0 => -1,
        :@fret1 => -1,
        :@fret2 => -1,
        :@fret3 => -1,
        :@fret4 => 5,
        :@fret5 => 5
      }
    ]

    @events = [
      {
        :@time => 10.0,
        :@code => 'B0'
      }
    ]

    @fret_hand_mutes = []

    @anchors = [
      {
        :@time => 275.931,
        :@fret => 13,
        :@width => 4.000
      }
    ]

    @hand_shapes = [
      {
        :@chordId => 0,
        :@endTime => 5.728,
        :@startTime => 5.672
      }
    ]
  end

  def fraction_of_bar(duration)
    # this name is a bit confusing
    1.0 / (2**Integer(I_DURATIONS[duration])) / @signature
  end

  def bar2time(barfraction)
    t = @timerinter.read(barfraction)
    (t * 1000).round / 1000.0
  end

  def new_note(string, note)
    bb = []
    if note.bend
      note.bend[:points].each do |b|
        bb << {
          :@time => @time + b[:time] / 1000.0,
          :@step => b[:pitch_alteration] / 100.0
        } # unless b[:time] == 0 and b[:pitch_alteration] == 0
      end
    end

    if note.slide
    end

    if note.hammer_or_pull
    end

    if note.grace
    end

    # sustain
    # hammer / pull / hopo
    # slides

    n = {
      :@time => @time,
      # :@linkNext => 0,
      :@accent => note.accentuated ? 1 : 0,
      :@bend => bb.count > 0 ? 1 : 0,
      :@fret => note.fret,
      # :@hammerOn => 0,
      :@harmonic => note.harmonic != :none ? 1 : 0, # note.harmonic != :pinch
      # :@hopo => 0,
      :@ignore => 0,
      :@leftHand => -1,
      :@mute => note.type == :dead ? 1 : 0,
      :@palmMute => note.palm_mute ? 1 : 0,
      :@pluck => -1,
      # :@pullOff => 0,
      :@slap => -1,
      # :@slideTo => -1,
      :@string => string,
      # :@sustain => 1.130,
      :@tremolo => note.tremolo ? 1 : 0,
      :@harmonicPinch => note.harmonic == :pinch ? 1 : 0,
      :@pickDirection => 0,
      :@rightHand => -1,
      # :@slideUnpitchTo => -1,
      :@tap => 0,
      :@vibrato => 0,
      :bendValues => carray(:bendValue, bb)
    }
    n
  end

  def new_chord(notes)
    # TODO: bendValues
    notes.each { |n| n.delete(:bendValues) }

    {
      :@time => @time,
      # :@linkNext => 0,
      :@accent => (notes.any? { |n| n[:@accent] == 1 }) ? 1 : 0,
      # :@chordId => 12,
      # :@fretHandMute => 0,
      # :@highDensity => 0, # criterion needed here
      :@ignore => notes.any? { |n| n[:@ignore] == 1 } ? 1 : 0,
      :@palmMute => notes.all? { |n| n[:@palmMute] == 1 } ? 1 : 0,
      :@hopo => notes.any? { |n| n[:@hopo] == 1 } ? 1 : 0,
      # :@strum => "down", # in the beat actually
      :chordNote => notes
    }
  end

  def new_bar(bar_settings)
    if bar_settings.new_time_signature
      s = bar_settings.new_time_signature
      @signature = 4.0 * s[:numerator] / s[:denominator]
    end

    @start_reapeat_bar = @bar_idx if bar_settings.has_start_of_repeat

    @sections << {
      :@name => bar_settings.marker[:name],
      :@number => @sections.count + 1,
      :@startTime => bar2time(@measure)
    } if bar_settings.marker

    @ebeats << {
      :@time => bar2time(@measure),
      :@measure => (@measure + 1).to_i
    }
    1.upto(@signature - 1).each do |i|
      @ebeats << {
        :@time => bar2time(@measure + i / @signature),
        :@measure => -1
      }
    end
  end

  def new_tone_change(tone_name)
    idx = @tone_map.index(tone_name)
    unless idx
      idx = @tone_map.count
      @tone_map << tone_name
    end

    @tones << {
      :@id => idx,
      :@time => @time
    }
  end

  def new_gp_beat(beat)
    new_tone_change beat.text if beat.text

    ns = beat.strings.map do |string, note|
      string = @track.strings.count - string.to_i
      new_note string, note
    end
    @notes << ns[0] if ns.count == 1
    @chords << new_chord(ns) if ns.count > 1

    @previous_notes = ns
  end

  def run
    @measure = 0.0
    @bar_idx = 0 # to handle repeats

    u = @track.bars.zip(@song.bars_settings)
    while @bar_idx < u.count
      bar, bar_settings = u[@bar_idx]
      new_bar bar_settings

      measure_fraction = 0.0
      bar.voices[:lead].each do |beat|
        @time = bar2time(@measure + measure_fraction)

        new_gp_beat beat

        measure_fraction += fraction_of_bar(beat.duration)
      end

      if bar_settings.has_end_of_repeat
        @repeats_count = bar_settings.repeats_count unless @repeats_count
      end

      if bar_settings.has_end_of_repeat && @repeats_count > 0
        @bar_idx = @start_reapeat_bar - 1
        @repeats_count -= 1
      elsif bar_settings.has_end_of_repeat && @repeats_count == 0
        @repeats_count = nil
        @start_reapeat_bar = nil
      end

      @bar_idx += 1
      @measure += 1.0
    end
  end

  def xml
    {
      :@version       => 8,
      :title          => @song.title,
      :arrangement    => @track.name,
      :wavefilepath   => '',
      :part           => 1,
      :offset         => -10.000,
      :centOffset     => 0,
      :songLength     => 0.000,
      :internalName   => @internal_name,
      :songNameSort   => sortable_name(@song.title),
      :startBeat      => 0.000,
      :averageTempo   => @song.bpm,
      :tuning         => compute_tuning(@track.strings),
      :capo           => @track.capo,
      :artistName     => @song.artist,
      :artistNameSort => sortable_name(@song.artist),
      :albumName      => @song.album,
      :albumNameSort  => sortable_name(@song.album),
      :albumYear      => @song.copyright.to_i,
      :albumArt       => '',
      :crowdSpeed     => 1,
      :arrangementProperties => {
        :@represent         => 1,
        :@bonusArr          => 0,
        :@standardTuning    => 1,
        :@nonStandardChords => 0,
        :@barreChords       => 0,
        :@powerChords       => 0,
        :@dropDPower        => 0,
        :@openChords        => 0,
        :@fingerPicking     => 0,
        :@pickDirection     => 0,
        :@doubleStops       => 0,
        :@palmMutes         => 0,
        :@harmonics         => 0,
        :@pinchHarmonics    => 0,
        :@hopo              => 0,
        :@tremolo           => 0,
        :@slides            => 0,
        :@unpitchedSlides   => 0,
        :@bends             => 0,
        :@tapping           => 0,
        :@vibrato           => 0,
        :@fretHandMutes     => 0,
        :@slapPop           => 0,
        :@twoFingerPicking  => 0,
        :@fifthsAndOctaves  => 0,
        :@syncopation       => 0,
        :@bassPick          => 0,
        :@sustain           => 0,
        :@pathLead          => 1,
        :@pathRhythm        => 0,
        :@pathBass          => 0
      },
      :lastConversionDateTime => Time.now.strftime('%F %T'),

      :tone__A => @tone_map.count > 1 ? @tone_map[1] : '',
      :tone__B => @tone_map.count > 2 ? @tone_map[2] : '',
      :tone__C => @tone_map.count > 3 ? @tone_map[3] : '',
      :tone__D => @tone_map.count > 4 ? @tone_map[4] : '',
      :tone__Base => '', # TODO: default values
      :tone__Multiplayer => '',
      :tones => carray(:tone, @tones),

      :phrases => carray(:phrase, @phrases),
      :phraseIterations => carray(:phraseIteration, @phrase_iterations),
      :newLinkedDiffs => carray(:newLinkedDiff, []),
      :linkedDiffs => carray(:linkedDiff, []),
      :phraseProperties => carray(:phrase, []),
      :chordTemplates => carray(:chordTemplate, @chord_templates),
      :fretHandMuteTemplates => carray(:fretHandMuteTemplate, []),
      :ebeats => carray(:ebeat, @ebeats),
      :sections => carray(:section, @sections),
      :events => carray(:event, @events),

      :transcriptionTrack => { :@difficulty => -1 },

      :levels => carray(:level, [
        {
          :@difficulty => 0,
          :notes => carray(:note, @notes),
          :chords => carray(:chord, @chords),
          :fretHandMutes => carray(:fretHandMute, @fret_hand_mutes),
          :anchors => carray(:anchor, @anchors),
          :handShapes => carray(:handShape, @hand_shapes)
        }
      ])
    }
  end
end

## SCRIPT
debug = 1 unless ARGV[0]
ARGV[0] = '../test/tab.xml' unless ARGV[0]

ARGV[0] = File.realpath ARGV[0]
dir = File.dirname ARGV[0]

gpa_xml = REXML::Document.new File.new(ARGV[0], 'r')
score_url = gpa_xml.elements['track'].elements['scoreUrl'].text
# mp3_url = gpa_xml.elements['track'].elements['mp3Url'].text
sync = gpa_xml.elements['track'].elements['sync'].text

# TODO: if no sync data use bpm
tabsong = GuitarProParser.read_file(dir + File::SEPARATOR + score_url)

tabsong.tracks[0, 1].each do |track|
  builder = SngXmlBuilder.new tabsong, track, time_interpolator(sync)
  builder.run
  puts Gyoku.xml :song => builder.xml unless debug == 1
end

File: bin/gpa2xml.py
#!/usr/bin/env python

"""
Convert Go PlayAlong to Rocksmith

Usage:
    gpa2xml.py FILE
"""

from time import strftime
import re
import json

from gpx2xml import has_prop, get_prop, load_goplayalong
from xmlhelpers import json2xml, InlineContent


DURATIONS = {
    'Long': 16.0,
    'DoubleWhole': 8.0,
    'Whole': 4.0,
    'Half': 2.0,
    'Quarter': 1.0,
    'Eighth': 1.0 / 2.0,
    '16th': 1.0 / 4.0,
    '32nd': 1.0 / 8.0,
    '64th': 1.0 / 16.0,
    '128th': 1.0 / 32.0,
    '256th': 1.0 / 64.0,
}


def text_for_sort(text):
    x = re.sub(r'(a|an|the)(\s+)', '', text, flags=re.IGNORECASE)
    return x.capitalize()


CH_TEMPLATES = json.loads(open('../share/chords.database.json').read())
CH_TEMPLATES = CH_TEMPLATES['Static']['Chords']['Entries']


# TODO: make the logic clearer :)
def find_fingering(chord):
    f0 = [chord['@fret' + str(k)] for k in range(6)]
    mask = [x > -1 for x in f0]
    f = f0
    min_fret = min([x for x in f if x > -1])
    if min_fret > 0:
        f = [x - min_fret + 1 for x in f]

    w = None
    for c in CH_TEMPLATES:
        z = [x < 0 or x == y for x, y in zip(f, c['Frets'])]
        if all(z):
            w = c
            break

    if w:
        k = [-1 if not y else x for x, y in zip(c['Fingers'], mask)]
        for kk in range(6):
            chord['@finger' + str(kk)] = k[kk]
            # print chord
        # print f0, '=>', k
    # else:
    #     print '### Not found', f0


class SngBuilder:
    def __init__(self, song, track, timefun):
        self.song = song
        self.track = track
        self.timefun = timefun

    def clear(self):
        self.phrases = []
        self.phraseIterations = []
        self.chordTemplates = []
        self.notes = []
        self.chords = []
        self.handShapes = []
        self.ebeats = []
        self.events = []
        self.sections = []
        self.tones = []

        self.measure = 0
        self.bar_idx = 0
        self.time = self.timefun(0)
        self.quarters_per_bar = 0.0
        self.offset_in_measure = 0.0

        self.anchors = [{
            '@time': 10.000,
            '@fret': 2,
            '@width': 4.000
        }]

        # TODO: from sections, double bar and repeats
        self.phrases = [{
            '@disparity': 0,
            '@ignore': 0,
            '@maxDifficulty': 0,
            '@name': 'default',
            '@solo': 0
        }]

        self.phraseIterations = [{
            '@time': 0.000,
            '@phraseId': 0,
            '@variation': ''
        }]

    def json(self):
        score = self.song.Score

        internalName = filter(str.isalnum, score.Artist)
        internalName += filter(str.isalnum, score.Title)
        part = 1  # TODO

        centOffset = 0  # No cent offset in GPX ?

        songLength = self.ebeats[-1]['@time'] if len(self.ebeats) > 0 else 0
        offset = self.timefun.offset
        averageTempo = (len(self.ebeats) - 1) / (songLength + offset) * 60
        averageTempo = int(averageTempo * 1000) / 1000.0

        STANDARD_TUNING = [40, 45, 50, 55, 59, 64]
        tuning = get_prop(self.track, 'Tuning', STANDARD_TUNING)
        standardTuning = int(tuning == STANDARD_TUNING)

        tuning = [a - b for a, b in zip(tuning, STANDARD_TUNING)]
        tuning = {'@string' + str(k): tuning[k] for k in range(6)}

        def xany(s, prop):
            return int(any(n[prop] for n in s))

        # TODO: check for techniques...
        arrangementProperties = {
            '@barreChords': 0,
            '@bassPick': 0,
            '@bends': xany(self.notes, '@bend'),
            '@bonusArr': 0,
            '@doubleStops': 0,
            '@dropDPower': 0,
            '@fifthsAndOctaves': 0,
            '@fingerPicking': 0,
            '@fretHandMutes': xany(self.chords, '@fretHandMute'),
            '@harmonics': xany(self.notes, '@harmonic'),
            '@hopo': xany(self.chords + self.notes, '@hopo'),
            '@nonStandardChords': 0,
            '@openChords': 0,
            '@palmMutes': xany(self.chords + self.notes, '@palmMute'),
            '@pathBass': 0,
            '@pathLead': 1,
            '@pathRhythm': 0,
            '@pickDirection': 0,
            '@pinchHarmonics': xany(self.notes, '@harmonicPinch'),
            '@powerChords': 0,
            '@represent': 1,
            '@slapPop': int(any(n['@slap'] != -1 or n['@pluck'] != -1 for n in self.notes)),
            '@slides': int(any(n['@slideTo'] != -1 for n in self.notes)),
            '@standardTuning': standardTuning,
            '@sustain': xany(self.notes, '@sustain'),
            '@syncopation': 0,
            '@tapping': xany(self.notes, '@tap'),
            '@tremolo': xany(self.notes, '@tremolo'),
            '@twoFingerPicking': 0,
            '@unpitchedSlides': int(any(n['@slideUnpitchTo'] != -1 for n in self.notes)),
            '@vibrato': xany(self.notes, '@vibrato')
        }

        return {
            '@version': 8,
            'albumArt': internalName,
            'albumName': score.Album,
            'albumNameSort': text_for_sort(score.Album),
            'albumYear': score.Copyright,
            'arrangement': self.track.Name,
            'arrangementProperties': arrangementProperties,
            'artistName': score.Artist,
            'artistNameSort': text_for_sort(score.Artist),
            'averageTempo': averageTempo,
            'capo': get_prop(self.track, 'CapoFret', 0),
            'centOffset': centOffset,
            'chordTemplates': self.chordTemplates,
            'crowdSpeed': 1,
            'ebeats': self.ebeats,
            'events': self.events,
            'fretHandMuteTemplates': [],
            'internalName': internalName,
            'lastConversionDateTime': strftime('%F %T'),
            'levels': [{
                '@difficulty': 0,
                'anchors': self.anchors,
                'chords': self.chords,
                'fretHandMutes': [],
                'handShapes': self.handShapes,
                'notes': self.notes
            }],
            'linkedDiffs': [],
            'newLinkedDiffs': [],
            'offset': offset,
            'part': part,
            'phraseIterations': self.phraseIterations,
            'phraseProperties': [],
            'phrases': self.phrases,
            'sections': self.sections,
            'songLength': songLength,
            'songNameSort': text_for_sort(self.track.Name),
            'startBeat': 0.000,
            'title': score.Title,
            'tone_A': '',
            'tone_B': '',
            'tone_C': '',
            'tone_D': '',
            'tone_Base': '',
            'tone_Multiplayer': '',
            'tones': self.tones,
            'transcriptionTrack': {
                '@difficulty': -1,
                'anchors': [],
                'chords': [],
                'handShapes': [],
                'notes': []
            },
            'tuning': tuning,
            'wavefilepath': ''
        }

    def new_chord(self, beat, notes):
        chordTemplate = {
            '@chordName': '',  # TODO
            '@displayName': '',  # TODO
            '@finger0': -1, '@finger1': -1, '@finger2': -1,
            '@finger3': -1, '@finger4': -1, '@finger5': -1,
            '@fret0': -1, '@fret1': -1, '@fret2': -1,
            '@fret3': -1, '@fret4': -1, '@fret5': -1
        }
        for n in notes:
            string, fret = n['@string'], n['@fret']
            chordTemplate['@fret' + str(string)] = fret

        find_fingering(chordTemplate)

        # print chordTemplate
        for n in notes:
            n['@leftHand'] = chordTemplate['@finger' + str(n['@string'])]

        if 'Arpeggio' in beat:
            chordTemplate['@displayName'] += '_arp'

        if not chordTemplate in self.chordTemplates:
            chordId = len(self.chordTemplates)
            self.chordTemplates.append(chordTemplate)
        else:
            chordId = self.chordTemplates.index(chordTemplate)

        self.handShapes.append({
            '@chordId': chordId,
            '@endTime': self.time + self.current_beat_length * 0.90,
            '@startTime': self.time
        })

        lowest_fret = [n['@fret'] for n in notes if n['@fret'] > -1]
        if lowest_fret != []:
            lowest_fret = min(lowest_fret)
            self.anchors.append({
                '@time': self.time,
                '@fret': lowest_fret,
                '@width': 4.000
            })

        # TODO check any or all ?
        return {
            '@accent': int(any(n['@accent'] for n in notes)),
            '@chordId': chordId,
            '@fretHandMute': int(any(n['@mute'] for n in notes)),
            '@highDensity': 0,
            '@hopo': int(any(n['@hopo'] for n in notes)),
            '@ignore': int(any(n['@ignore'] for n in notes)),
            '@linkNext': int(any(n['@linkNext'] for n in notes)),
            '@palmMute': int(any(n['@palmMute'] for n in notes)),
            '@strum': get_prop(beat, 'Direction', 'Down').lower(),
            '@time': self.time,
            'chordNotes': InlineContent(notes)
        }

    def new_notes(self, beat, notes):
        ns = []
        for n in notes:
            note = self.song.Notes[n]
            # print note

            # if has_prop(note, 'Slide'):
                # print note.Properties.Property
                # print get_prop(note, 'Slide')

            # if has_prop(note, 'Bended'):
            #     print note.Properties.Property

            # TODO slide, bend, hopo, sustain
            # handle fingering already there

            harmonic = get_prop(note, 'HarmonicType')

            ns.append({
                '@accent': int('Accent' in note),
                '@bend': 0,
                '@fret': get_prop(note, 'Fret'),
                '@hammerOn': 0,
                '@harmonic': int(harmonic == 'Artificial'),
                '@harmonicPinch': int(harmonic == 'Pinch'),
                '@hopo': 0,
                '@ignore': 0,
                '@leftHand': -1,
                '@linkNext': int('Tie' in note and note.Tie['@origin']),
                '@mute': int(has_prop(note, 'Muted')),
                '@palmMute': int(has_prop(note, 'PalmMuted')),
                '@pickDirection': 0,
                '@pluck': -1,  # int(has_prop(beat, 'Popped')),
                '@pullOff': 0,
                '@rightHand': -1,
                '@slap': -1,  # int(has_prop(beat, 'Slapped')),
                '@slideTo': -1,
                '@slideUnpitchTo': -1,
                '@string': get_prop(note, 'String'),
                '@sustain': 0.0,
                '@tap': int(has_prop(note, 'Tapped')),
                '@time': self.time,
                '@tremolo': int('Tremolo' in beat),
                '@vibrato': int('Vibrato' in note),
                'bendValues': []
            })

        if len(ns) > 1:
            self.chords.append(self.new_chord(beat, ns))
        else:
            self.notes += ns

    def new_beat(self, beat):
        rhythm = self.song.Rhythms[beat.Rhythm['@ref']]
        inc = DURATIONS[rhythm.NoteValue] / self.quarters_per_bar

        if 'PrimaryTuplet' in rhythm:
            tuplet = rhythm['PrimaryTuplet']
            inc *= tuplet['@den'] / tuplet['@num']

        if 'AugmentationDot' in rhythm:
            inc *= 1.5

        if 'GraceNotes' in beat and beat.GraceNotes == 'BeforeBeat':
            self.offset_in_measure -= inc
            self.time = self.timefun(self.measure + self.offset_in_measure)

        # TODO: hmmm
        self.current_beat_length = self.timefun(self.measure + self.offset_in_measure + inc) - self.time

        if 'FreeText' in beat:
            self.tones.append({
                '@id': 0,  # TODO
                '@time': self.time
            })

        # TODO: handle current hand position
        if 'Notes' in beat:
            if type(beat.Notes) is not list:
                beat.Notes = [beat.Notes]

            self.new_notes(beat, beat.Notes)

        self.offset_in_measure += inc
        self.time = self.timefun(self.measure + self.offset_in_measure)

        # Time stays the same but next note duration is diminished by inc
        if 'GraceNotes' in beat and beat.GraceNotes == 'OnBeat':
            self.offset_in_measure -= inc

    def new_bar(self, bar, bar_settings):
        num, den = map(int, bar_settings.Time.split('/'))
        self.quarters_per_bar = 4.0 * num / den

        if 'Repeat' in bar_settings and bar_settings.Repeat['@start']:
            self.start_repeat_bar = self.bar_idx

        # Need to update phrases too
        # if 'Section' in bar_settings:
        #     self.sections.append({
        #         '@name': bar_settings.Section.Text,
        #         '@number': len(self.sections),  # TODO
        #         '@startTime': self.timefun(self.measure)
        #     })

        for i in range(int(self.quarters_per_bar)):
            self.ebeats.append({
                '@time': self.timefun(self.measure + i / self.quarters_per_bar),
                '@measure': self.measure + 1 if i == 0 else -1
            })

        self.offset_in_measure = 0.0
        # TODO: other voices
        for b in self.song.Voices[bar.Voices[0]].Beats:
            beat = self.song.Beats[b]
            self.new_beat(beat)

        if 'Repeat' in bar_settings and bar_settings.Repeat['@end']:
            if not hasattr(self, 'repeats_count'):
                self.repeats_count = bar_settings.Repeat['@count']

            if self.repeats_count > 1:
                self.bar_idx = self.start_repeat_bar - 1
                self.repeats_count -= 1
            else:
                del self.repeats_count
                del self.start_repeat_bar

    def run(self):
        self.clear()
        while self.bar_idx < len(self.song.MasterBars):
            bar_settings = self.song.MasterBars[self.bar_idx]
            bar = self.song.Bars[bar_settings.Bars[self.track['@id']]]

            # TODO: beat value
            # gp.MasterTrack['Automations']['Automation']['Type'] == 'Tempo'
            self.new_bar(bar, bar_settings)

            self.bar_idx += 1  # no repeats -> get position in score
            self.measure += 1  # counting repeats -> compute time

        return self.json()


if __name__ == '__main__':
    from docopt import docopt
    from xml.dom.minidom import parseString

    args = docopt(__doc__)

    gp, mp3, sync = load_goplayalong(args['FILE'])

    # TODO: other tracks
    sng = SngBuilder(gp, gp.Tracks[0], sync)

    x = json2xml('song', sng.run())
    print parseString(x).toprettyxml()
